src/box/lua/iproto.c

```diff
@@ -268,16 +299,124 @@ push_iproto_protocol_features(struct lua_State *L)
 					  lengthof(features));
 }
 
+/**
+ * Encodes Lua table packet part to MsgPack using IPROTO key translation table:
+ * uses @a ibuf to store the result and saves the result length in @a mp_len.
+ */
+static const char *
+encode_lua_table_packet_part_to_mp(struct lua_State *L, int idx, size_t *mp_len)
+{
+	struct region *gc = &fiber()->gc;
+	struct mpstream stream;
+	mpstream_init(&stream, gc, region_reserve_cb, region_alloc_cb,
+		      luamp_error, L);
+	size_t used = region_used(gc);
+	luamp_encode_with_translation(L, luaL_msgpack_default, &stream, idx,
+				      iproto_key_translation);
+	mpstream_flush(&stream);
+	*mp_len = region_used(gc) - used;
+
+	const char *mp = region_join(gc, *mp_len);
+	if (mp == NULL) {
+		diag_set(OutOfMemory, *mp_len, "mp", "MsgPack data");
+		region_truncate(gc, used);
+	}
+	return mp;
+}
+
+/**
+ * Converts a packet part argument to MsgPack: if the argument is a string, then
+ * no conversion is needed — otherwise the argument must be a Lua table, which
+ * is encoded using `encode_lua_table_packet_part_to_mp`.
+ */
+static const char *
+convert_packet_part_arg_to_mp(struct lua_State *L, int idx, size_t *mp_len)
+{
+	int packet_part_type = lua_type(L, idx);
+	if (packet_part_type != LUA_TSTRING && packet_part_type != LUA_TTABLE) {
+		const char *arg = tt_sprintf("expected %d argument to be a "
+					     "string or table", idx);
+		diag_set(LuajitError, arg);
+		return NULL;
```

Comment:
Mixing Lua and C errors in the same function doesn't look good. Why not just call `luaL_error` here?

Reply:
I agree: I just didn't know about handy way to create a formatted error using Lua API — then I found out it's `lua_pushfstring` + `lua_error`. Fixed.

Reply:
Because otherwise we will leak region memory, when parsing the body parameter.

Reply:
Then let's check function arguments at the top level before allocating anything from the region.

Reply:
Good point, fixed — now `process_packet_part_arg` never fails, the code got simpler.